File: scripts/pabi_traj.py
#!/usr/bin/python

from std_srvs.srv import Trigger
from roboy_communication_middleware.msg import JointStatus
import os
import sys
import time

import rospy
import json

global angles
global times
global sub

def start_record(req):

	angles = []
	times = []
	trajectory = {}
	response = {}

	def callback(data):
		angles.append(data.relAngles)
		times.append(time.time())

	start = time.time()
	rospy.Subscriber("/roboy/middleware/JointStatus", JointStatus, callback)

	while True:
		print('recording')
		response["message"] = "Recording PaBi trajectory"

	response["success"] = True
	return json.dumps(response)

def stop_record(req):
	response = {}
	sub.unregister()
	# print('saving the data')
	# for i in range(len(times)):
	# 	trajectory[start - times[i]] = angles
	# with open('trajectory', 'w') as outfile:
	# 	json.dump(trajectory, outfile)

	response["success"] = True	
	response["message"] = "Recording PaBi trajectory was successful"

	angles = []
	times = []

	return json.dumps(response)

def replay(req):
	response = {}
	response["success"] = True
	response["message"] = "Replaying PaBi trajectory was successful"
	return json.dumps(response)


if __name__ == '__main__':
	angles = []
	times = []
	rospy.init_node('roboy_pabi_trajectories')

	rospy.Service('/roboy/control/pabi/trajectory/record/start', Trigger, start_record)
	rospy.Service('/roboy/control/pabi/trajectory/record/stop', Trigger, stop_record)
	rospy.Service('/roboy/control/pabi/trajectory/replay', Trigger, replay)
	
	print "/roboy/control/pabi/trajectory/ is ready"

	rospy.spin()

File: src/pabi_traj.cpp
#include <ros/ros.h>
#include <std_srvs/Trigger.h>
#include <roboy_communication_control/StartRecordTrajectory.h>
#include <roboy_communication_control/StopRecordTrajectory.h>

#include <fstream>
#include <iostream>
#include <json.hpp>

using json = nlohmann::json;

class TrajectoriesManager
{
	private:
	bool recording;
	json trajectory;

	public:
		TrajectoriesManager(){
			this->recording = false;
			// this->trajectory = "";
		}
		~TrajectoriesManager();



	bool startRecord(roboy_communication_control::StartRecordTrajectory::Request &req,
	                     roboy_communication_control::StartRecordTrajectory::Response &res) {

	      this->recording = true;
	      ROS_INFO_STREAM("recording start");

	      // json j = "{ \"happy\": true, \"pi\": 3.141 }"_json;
	      int j = 0;
	      while(this->recording)
	      {
			this->trajectory[j] = "fuck";
			j++;
	      }
	      ROS_INFO_STREAM("recording done");
	      res.success = true;
	      return true;
	  }

	bool stopRecord(roboy_communication_control::StopRecordTrajectory::Request &req,
	                     roboy_communication_control::StopRecordTrajectory::Response &res) {
	      this->recording = false;
	      res.trajectory = this->trajectory.dump();
	      this->trajectory = "";
	      return true;
	  }
	
};

int main(int argc, char **argv) {

	ros::init(argc, argv, "talker");

	ros::NodeHandle n;

	auto tm = new TrajectoriesManager();

	auto startRecordSrv = n.advertiseService("/roboy/control/pabi/trajectory/record/start", &TrajectoriesManager::startRecord, tm);
	auto stopRecordSrv = n.advertiseService("/roboy/control/pabi/trajectory/record/stop", &TrajectoriesManager::stopRecord, tm);

	// ros::Subscriber jointStatusSub = n.subscribe("/roboy/middleware/JointStatus", 1000, &MyoMotor::moveMotorCallback, &myo_control);

	ros::AsyncSpinner spinner (0);
	while (ros::ok())
	{
	 spinner.start();
	}
	spinner.stop();
}

